packages/test-support/addon/jsonapi-factory.js

```diff
@@ -1,20 +1,37 @@
 import DAGMap from 'dag-map';
 
 let idGenerator = 0;
+const DELIM = '::';
 
 export default class JSONAPIFactory {
-  constructor() {
+  constructor(sourceId, packageName) {
     this.data = [];
+    this.sourceId = sourceId;
+    this.packageName = packageName;
   }
 
-  addResource(type, id=null) {
-    if (id == null) {
-      id = idGenerator++;
+  addResource(type, _id=null) {
+    let id = _id;
+    if (type === 'card-definitions' && this.sourceId != null && this.packageName != null) {
```

Comment:
The fact that JSONAPIFactory needs to know about the composite id scheme and needs to get it correct is a problem. 

The factory is a convenience DSL only, people can author their schema.js in pure json too (and that will become the more common thing to do as soon as we have tools for manipulating the schema, instead of hand writing it). But this means that people will be free to author invalid schema.js files because they don't follow the composite id format, which is not a thing I want to expose to users.

And while we can easily detect invalid ids and bully users into typing out correct ones, that's unnecessary when we definitely have enough information to do the right thing automatically.

It would be better if we could forbid users from picking IDs at all in here. But to express that in json:api, we would need to use json:api operations format, rather than a straight-up json:api document. That may be the way to go. Alternatively, we can just state that the ids you put into schema.js are local to that file only and you don't get to expect them to mean anything other than expressing relationships inside that file.

Reply:
what about all the existing uses of `JSONAPIFactory` for arbitrary test setup? should I refactor all that to use just plain objects? or should we provide a means that you can assert an ID in the resource creation for the time being, until we actually start using the new schemas in earnest?